src/base/ActionsRouter.sol

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.24;
+
+import {IPoolManager} from "@uniswap/v4-core/src/interfaces/IPoolManager.sol";
+import {Currency} from "@uniswap/v4-core/src/types/Currency.sol";
+import {TransientStateLibrary} from "@uniswap/v4-core/src/libraries/TransientStateLibrary.sol";
+
+import {SafeTransferLib} from "solmate/utils/SafeTransferLib.sol";
+
+import {SafeCallback} from "./SafeCallback.sol";
+import {ReentrancyLock} from "./ReentrancyLock.sol";
+import {Actions} from "../libraries/Actions.sol";
+import {Locker} from "../libraries/Locker.sol";
+
+/// @notice Abstract contract for performing a combination of actions on Uniswap v4.
+/// @dev If an inheriting contract does not want to support all commands, they should simply revert with the
+/// UnsupportedAction error in their overriding function
+/// If additional actions are required, define new unique action constants, and implement the logic for handling them
+/// in the `_handleAdditionalActions{1/2}` functions
+/// The _msgSender function returns the msg.sender from the call to executeActions. This allows actions to know which
+/// address instructed the action to be requested. The msg.sender within the unlockCallback with be the v4 PoolManager.
+abstract contract BaseActionsRouter is SafeCallback, ReentrancyLock {
+    using TransientStateLibrary for IPoolManager;
+    using SafeTransferLib for address;
+
+    /// @notice emitted when different numbers of parameters and actions are provided
+    error LengthMismatch();
+
+    /// @notice emitted when an inheriting contract does not support an action
+    error UnsupportedAction(uint256 action);
+
+    constructor(IPoolManager poolManager) SafeCallback(poolManager) {}
+
+    /// @notice internal function that triggers the execution of a set of actions on v4
+    function _executeActions(bytes calldata params) internal isNotLocked {
+        poolManager.unlock(params);
+    }
+
+    /// @notice function that is called by the PoolManager through the SafeCallback.unlockCallback
+    function _unlockCallback(bytes calldata data) internal override returns (bytes memory) {
+        // TODO decode in calldata directly for gas
+        (uint256[] memory actions, bytes[] memory params) = abi.decode(data, (uint256[], bytes[]));
+
+        uint256 numActions = actions.length;
+        if (numActions != params.length) revert LengthMismatch();
+
+        for (uint256 actionIndex = 0; actionIndex < numActions; actionIndex++) {
+            uint256 action = actions[actionIndex];
+
+            // separate actions on a pool, and settling deltas
+            if (action < 0x10) {
```

Comment:
The one bummer is that this will be slightly less optimized since posm will only use INCREASE, DECREASE, TAKE basically.. and so the ordering of checks would just be different for different use cases

Reply:
itll be using settle too right? So settle and take are optimally positioned, its just liduidity commands that are put 1 lower. We can look at gas differences but I'd expect it to be <50 gas difference